File: aidoorlock.c
/*
 * Copyright 2010-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/**
 * @file subscribe_publish_sample.c
 * @brief simple MQTT publish and subscribe on the same topic
 *
 * This example takes the parameters from the aws_iot_config.h file and establishes a connection to the AWS IoT MQTT Platform.
 * It subscribes and publishes to the same topic - TOPIC_LOCKS_CMD
 *
 * If all the certs are correct, you should see the messages received by the application in a loop.
 *
 * The application takes in the certificate path, host name , port and the number of times the publish should happen.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>

#include <signal.h>
#include <memory.h>
#include <sys/time.h>
#include <limits.h>

#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <netinet/in.h>
#include <net/if.h>
#include <arpa/inet.h>

#include "aws_iot_log.h"
#include "aws_iot_version.h"
#include "aws_iot_mqtt_interface.h"
#include "aws_iot_config.h"

#include "constants.h"

static char passcode[5] = "0000";

void getSelfIP(char *iface, char* ip) {
	int fd;
	struct ifreq ifr;

	fd = socket(AF_INET, SOCK_DGRAM, 0);
	ifr.ifr_addr.sa_family = AF_INET;

	strncpy(ifr.ifr_name, iface, IFNAMSIZ-1);
	ioctl(fd, SIOCGIFADDR, &ifr);
	//printf("%s\n", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr));
	sprintf(ip, "%s\n", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr));

	close(fd);
}

int onStartup() {
	INFO("Announcing startup");
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_READY "\" &");

	//publish own IP address to topic "locks/ip"
	IoT_Error_t rc = NONE_ERROR;
	MQTTMessageParams Msg = MQTTMessageParamsDefault;
	Msg.qos = QOS_0;
	char cPayload[100];

	MQTTPublishParams Params = MQTTPublishParamsDefault;
	Params.pTopic = "locks/ip";

	char wlan0_ip[50] = "";
	getSelfIP("wlan0", wlan0_ip);
	sprintf(cPayload, "wlan0 IP address is %s", wlan0_ip);
	Msg.pPayload = (void *) cPayload;
	Msg.PayloadLen = strlen(cPayload) + 1;

	Params.MessageParams = Msg;
	rc = aws_iot_mqtt_publish(&Params);
	if(rc != NONE_ERROR) {
		printf("Error publishing IP address to 'locks/ip'\n");
	}

	printf("wlan0 IP address is %s", wlan0_ip);

	char eth0_ip[50] = "";
	getSelfIP("eth0", eth0_ip);
	sprintf(cPayload, "eth0 IP address is %s", eth0_ip);
	Msg.pPayload = (void *) cPayload;
	Msg.PayloadLen = strlen(cPayload) + 1;

	Params.MessageParams = Msg;
	rc = aws_iot_mqtt_publish(&Params);
	if(rc != NONE_ERROR) {
		printf("Error publishing IP address to 'locks/ip'\n");
	}

	printf("eth0 IP address is %s", eth0_ip);
}

int cmdHandlerCapturePhoto(MQTTCallbackParams params) {
	INFO("Capture Photo");
	//important to add '&' at end of command so that the MQTT subscription of this program does not get blocked (and timed out)
	system("sh `pwd`/scripts/capture.sh \"" POLLY_PROMPT_LOOK_AT_CAMERA "\" \"" POLLY_PROMPT_WAIT_A_MOMENT "\" &");
}

int cmdHandlerFrFailure(MQTTCallbackParams params) {
	INFO("Facial Recognition Failed");
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_FR_FAILURE "\" &");
}

int cmdHandlerUpdatePasscode(MQTTCallbackParams params) {
	INFO("Update Passcode");
	char *pPasscode = ((char*)params.MessageParams.pPayload)+strlen(CMD_UPDATE_PASSCODE)+1;
	strncpy(passcode, pPasscode, 4);
	INFO("New passcode is %s", passcode);
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_SENDING_SMS "\" &");
}

int cmdHandlerAskSecret(MQTTCallbackParams params) {
	INFO("Ask Secret");
	char command[1000];
	bzero(command, sizeof(command));
	sprintf(command,
		"%s %s %s %s %s",
		"sh `pwd`/scripts/passcode.sh ",
		passcode,
		" \"" POLLY_PROMPT_ASK_SECRET "\"",
		" \"" POLLY_PROMPT_ALLOW_ACCESS "\"",
		" \"" POLLY_PROMPT_DENY_ACCESS "\" &");
	system(command);
	//system("sh `pwd`/scripts/passcode.sh \"" POLLY_PROMPT_ASK_SECRET "\" &");
}

int cmdHandlerAllowAccess(MQTTCallbackParams params) {
	INFO("Allow Access");
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_ALLOW_ACCESS "\" &");
}

int cmdHandlerDenyAccess(MQTTCallbackParams params) {
	INFO("Deny Access");
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_DENY_ACCESS "\" &");
}

int cmdHandlerSMSFailed(MQTTCallbackParams params) {
	INFO("SMS Failed");
	system("python `pwd`/scripts/speak.py \"" POLLY_PROMPT_SMS_FAILED "\" &");
}

int MQTTcallbackHandler(MQTTCallbackParams params) {

	INFO("Subscribe callback");
	INFO("%.*s\t%.*s",
			(int)params.TopicNameLen, params.pTopicName,
			(int)params.MessageParams.PayloadLen, (char*)params.MessageParams.pPayload);
	if(strncmp((char*)params.MessageParams.pPayload, CMD_CAPTURE_PHOTO, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerCapturePhoto(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_FR_FAILURE, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerFrFailure(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_UPDATE_PASSCODE, strlen(CMD_UPDATE_PASSCODE)) == 0) {
		cmdHandlerUpdatePasscode(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_ASK_SECRET, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerAskSecret(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_ALLOW_ACCESS, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerAllowAccess(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_DENY_ACCESS, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerDenyAccess(params);
	}
	else if(strncmp((char*)params.MessageParams.pPayload, CMD_SMS_FAILED, (int)params.MessageParams.PayloadLen) == 0) {
		cmdHandlerSMSFailed(params);
	}

	return 0;
}

void disconnectCallbackHandler(void) {
	WARN("MQTT Disconnect");
	IoT_Error_t rc = NONE_ERROR;
	if(aws_iot_is_autoreconnect_enabled()){
		INFO("Auto Reconnect is enabled, Reconnecting attempt will start now");
	}else{
		WARN("Auto Reconnect not enabled. Starting manual reconnect...");
		rc = aws_iot_mqtt_attempt_reconnect();
		if(RECONNECT_SUCCESSFUL == rc){
			WARN("Manual Reconnect Successful");
		}else{
			WARN("Manual Reconnect Failed - %d", rc);
		}
	}
}

/**
 * @brief Default cert location
 */
char certDirectory[PATH_MAX + 1] = "../../certs";

/**
 * @brief Default MQTT HOST URL is pulled from the aws_iot_config.h
 */
char HostAddress[255] = AWS_IOT_MQTT_HOST;

/**
 * @brief Default MQTT port is pulled from the aws_iot_config.h
 */
uint32_t port = AWS_IOT_MQTT_PORT;

/**
 * @brief This parameter will avoid infinite loop of publish and exit the program after certain number of publishes
 */
uint32_t publishCount = 0;

void parseInputArgsForConnectParams(int argc, char** argv) {
	int opt;

	while (-1 != (opt = getopt(argc, argv, "h:p:c:x:"))) {
		switch (opt) {
		case 'h':
			strcpy(HostAddress, optarg);
			DEBUG("Host %s", optarg);
			break;
		case 'p':
			port = atoi(optarg);
			DEBUG("arg %s", optarg);
			break;
		case 'c':
			strcpy(certDirectory, optarg);
			DEBUG("cert root directory %s", optarg);
			break;
		case 'x':
			publishCount = atoi(optarg);
			DEBUG("publish %s times\n", optarg);
			break;
		case '?':
			if (optopt == 'c') {
				ERROR("Option -%c requires an argument.", optopt);
			} else if (isprint(optopt)) {
				WARN("Unknown option `-%c'.", optopt);
			} else {
				WARN("Unknown option character `\\x%x'.", optopt);
			}
			break;
		default:
			ERROR("Error in command line argument parsing");
			break;
		}
	}

}

int main(int argc, char** argv) {
	IoT_Error_t rc = NONE_ERROR;
	int32_t i = 0;
	bool infinitePublishFlag = true;

	char rootCA[PATH_MAX + 1];
	char clientCRT[PATH_MAX + 1];
	char clientKey[PATH_MAX + 1];
	char CurrentWD[PATH_MAX + 1];
	char cafileName[] = AWS_IOT_ROOT_CA_FILENAME;
	char clientCRTName[] = AWS_IOT_CERTIFICATE_FILENAME;
	char clientKeyName[] = AWS_IOT_PRIVATE_KEY_FILENAME;

	parseInputArgsForConnectParams(argc, argv);

	INFO("\nAWS IoT SDK Version %d.%d.%d-%s\n", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH, VERSION_TAG);

	getcwd(CurrentWD, sizeof(CurrentWD));
	sprintf(rootCA, "%s/%s/%s", CurrentWD, certDirectory, cafileName);
	sprintf(clientCRT, "%s/%s/%s", CurrentWD, certDirectory, clientCRTName);
	sprintf(clientKey, "%s/%s/%s", CurrentWD, certDirectory, clientKeyName);

	DEBUG("rootCA %s", rootCA);
	DEBUG("clientCRT %s", clientCRT);
	DEBUG("clientKey %s", clientKey);

	MQTTConnectParams connectParams = MQTTConnectParamsDefault;

	connectParams.KeepAliveInterval_sec = 10;
	connectParams.isCleansession = true;
	connectParams.MQTTVersion = MQTT_3_1_1;
	connectParams.pClientID = "CSDK-test-device";
	connectParams.pHostURL = HostAddress;
	connectParams.port = port;
	connectParams.isWillMsgPresent = false;
	connectParams.pRootCALocation = rootCA;
	connectParams.pDeviceCertLocation = clientCRT;
	connectParams.pDevicePrivateKeyLocation = clientKey;
	connectParams.mqttCommandTimeout_ms = 2000;
	connectParams.tlsHandshakeTimeout_ms = 5000;
	connectParams.isSSLHostnameVerify = true; // ensure this is set to true for production
	connectParams.disconnectHandler = disconnectCallbackHandler;

	INFO("Connecting...");
	rc = aws_iot_mqtt_connect(&connectParams);
	if (NONE_ERROR != rc) {
		ERROR("Error(%d) connecting to %s:%d", rc, connectParams.pHostURL, connectParams.port);
		goto EXIT_CLEANUP;
	}
	/*
	 * Enable Auto Reconnect functionality. Minimum and Maximum time of Exponential backoff are set in aws_iot_config.h
	 *  #AWS_IOT_MQTT_MIN_RECONNECT_WAIT_INTERVAL
	 *  #AWS_IOT_MQTT_MAX_RECONNECT_WAIT_INTERVAL
	 */
	rc = aws_iot_mqtt_autoreconnect_set_status(true);
	if (NONE_ERROR != rc) {
		ERROR("Unable to set Auto Reconnect to true - %d", rc);
		goto EXIT_CLEANUP;
	}

	MQTTSubscribeParams subParams = MQTTSubscribeParamsDefault;
	subParams.mHandler = MQTTcallbackHandler;
	subParams.pTopic = TOPIC_LOCKS_CMD;
	subParams.qos = QOS_0;

	if (NONE_ERROR == rc) {
		INFO("Subscribing...");
		rc = aws_iot_mqtt_subscribe(&subParams);
		if (NONE_ERROR != rc) {
			ERROR("Error subscribing");
			goto EXIT_CLEANUP;
		}
	}

	MQTTMessageParams Msg = MQTTMessageParamsDefault;
	Msg.qos = QOS_0;
	char cPayload[100];
	sprintf(cPayload, "%s : %d ", "hello from SDK", i);
	Msg.pPayload = (void *) cPayload;

	MQTTPublishParams Params = MQTTPublishParamsDefault;
	Params.pTopic = TOPIC_LOCKS_CMD;

	if (publishCount != 0) {
		infinitePublishFlag = false;
	}

	//Announce readiness
	onStartup();

	while ((NETWORK_ATTEMPTING_RECONNECT == rc || RECONNECT_SUCCESSFUL == rc || NONE_ERROR == rc)
			&& (publishCount > 0 || infinitePublishFlag)) {

		//Max time the yield function will wait for read messages
		rc = aws_iot_mqtt_yield(100);
		if(NETWORK_ATTEMPTING_RECONNECT == rc){
			//INFO("-->sleep");
			sleep(1);
			// If the client is attempting to reconnect we will skip the rest of the loop.
			continue;
		}
		//INFO("-->sleep");
		sleep(1);
/*
		sprintf(cPayload, "%s : %d ", "hello from SDK", i++);
		Msg.PayloadLen = strlen(cPayload) + 1;
		Params.MessageParams = Msg;
		rc = aws_iot_mqtt_publish(&Params);
		if (publishCount > 0) {
			publishCount--;
		}
*/
	}

EXIT_CLEANUP:
	if (NONE_ERROR != rc) {
		ERROR("An error occurred, exiting.\n");
	} else {
		INFO("Exiting without errors.\n");
	}

	return rc;
}

File: aidoorlock.py
#!/usr/bin/env python3
#
# Copyright 2017 Amazon.com, Inc. and its affiliates. All Rights Reserved.
#
# Licensed under the Amazon Software License (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#   http://aws.amazon.com/asl/
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

"""
aidoorlock.py - AWS IoT connected door lock device program.

Connects to AWS IoT Core via MQTT, subscribes to the 'locks/commands' topic,
and dispatches commands to helper scripts for camera capture, speech synthesis,
and passcode verification.
"""

import json
import logging
import os
import signal
import socket
import subprocess
import sys
import threading
import time

from awscrt import io, mqtt
from awsiot import mqtt_connection_builder

# ---------------------------------------------------------------------------
# Constants - MQTT Topics
# ---------------------------------------------------------------------------
TOPIC_LOCKS_CMD = "locks/commands"
TOPIC_LOCKS_IP = "locks/ip"

# ---------------------------------------------------------------------------
# Constants - Command Strings
# ---------------------------------------------------------------------------
CMD_CAPTURE_PHOTO = "CAPTURE PHOTO"
CMD_FR_FAILURE = "FACIAL VERIFICATION FAILED"
CMD_UPDATE_PASSCODE = "UPDATE PASSCODE"
CMD_ASK_SECRET = "ASK SECRET"
CMD_ALLOW_ACCESS = "ALLOW ACCESS"
CMD_DENY_ACCESS = "DENY ACCESS"
CMD_SMS_FAILED = "SMS FAILED"

# ---------------------------------------------------------------------------
# Constants - Polly SSML Prompts
# ---------------------------------------------------------------------------
POLLY_PROMPT_READY = "<speak>Doorlock is ready</speak>"
POLLY_PROMPT_LOOK_AT_CAMERA = (
    "<speak>Hello! Please look at the camera. Remember to remove your glasses "
    "if you are wearing any.<break time='1s'/> Three. Two. One. And click!</speak>"
)
POLLY_PROMPT_WAIT_A_MOMENT = "<speak>Please wait a moment.</speak>"
POLLY_PROMPT_FR_FAILURE = (
    "<speak>Nope. I could not find you on the expected guest list. Go away!</speak>"
)
POLLY_PROMPT_ASK_SECRET = (
    "<speak>A passcode has been sent to your phone, please read it out aloud.</speak>"
)
POLLY_PROMPT_ALLOW_ACCESS = "<speak>Welcome, you may enter now!</speak>"
POLLY_PROMPT_DENY_ACCESS = (
    "<speak><prosody volume='+20dB'>Nope, that passcode is incorrect. "
    "I have alerted the police and the landlord. Go away, run for your life!</prosody></speak>"
)
POLLY_PROMPT_SENDING_SMS = (
    "<speak>You will receive an SMS with passcode shortly, please standby.</speak>"
)
POLLY_PROMPT_SMS_FAILED = (
    "<speak>I am sorry, I was unable to send SMS with passcode to your "
    "registered phone number.</speak>"
)

# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger("aidoorlock")

# ---------------------------------------------------------------------------
# Global state
# ---------------------------------------------------------------------------
passcode = "0000"
passcode_lock = threading.Lock()
base_dir = os.path.dirname(os.path.abspath(__file__))
shutdown_event = threading.Event()


def load_config(config_path=None):
    """Load configuration from config.json."""
    if config_path is None:
        config_path = os.path.join(base_dir, "config.json")
    with open(config_path, "r") as f:
        return json.load(f)


def get_self_ip(interface):
    """Get the IP address of a network interface (cross-platform)."""
    if sys.platform == "linux":
        try:
            import fcntl
            import struct
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            ip = socket.inet_ntoa(
                fcntl.ioctl(
                    s.fileno(),
                    0x8915,  # SIOCGIFADDR
                    struct.pack("256s", interface.encode("utf-8")[:15]),
                )[20:24]
            )
            s.close()
            return ip
        except Exception:
            return None
    elif sys.platform == "darwin":
        # macOS: use socket.getaddrinfo as a fallback
        try:
            # Map common Linux interface names to a hostname lookup
            # On macOS, we resolve the hostname to get a routable IP
            hostname = socket.gethostname()
            addrs = socket.getaddrinfo(hostname, None, socket.AF_INET)
            for addr in addrs:
                ip = addr[4][0]
                if ip and not ip.startswith("127."):
                    return ip
            return None
        except Exception:
            return None
    else:
        # Unsupported platform
        return None


def run_speak(text, region):
    """Run speak.py in a background subprocess (non-blocking)."""
    script = os.path.join(base_dir, "scripts", "speak.py")
    subprocess.Popen(
        ["python3", script, text, region],
        cwd=base_dir,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def run_capture(prompt1, prompt2, region, bucket):
    """Run capture.sh in a background subprocess (non-blocking)."""
    script = os.path.join(base_dir, "scripts", "capture.sh")
    subprocess.Popen(
        ["sh", script, prompt1, prompt2, region, bucket],
        cwd=base_dir,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def run_passcode(code, ask_prompt, allow_prompt, deny_prompt, region):
    """Run passcode.sh in a background subprocess (non-blocking)."""
    script = os.path.join(base_dir, "scripts", "passcode.sh")
    subprocess.Popen(
        ["sh", script, code, ask_prompt, allow_prompt, deny_prompt, region],
        cwd=base_dir,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


# ---------------------------------------------------------------------------
# Command Handlers
# ---------------------------------------------------------------------------

def cmd_handler_capture_photo(config):
    """Handle CAPTURE PHOTO command."""
    logger.info("Capture Photo")
    run_capture(
        POLLY_PROMPT_LOOK_AT_CAMERA,
        POLLY_PROMPT_WAIT_A_MOMENT,
        config["host_region"],
        config["s3_bucket_name"],
    )


def cmd_handler_fr_failure(config):
    """Handle FACIAL VERIFICATION FAILED command."""
    logger.info("Facial Recognition Failed")
    run_speak(POLLY_PROMPT_FR_FAILURE, config["host_region"])


def cmd_handler_update_passcode(payload, config):
    """Handle UPDATE PASSCODE command - extract 4-digit passcode."""
    global passcode
    logger.info("Update Passcode")
    # Payload format: "UPDATE PASSCODE XXXX"
    new_passcode = payload[len(CMD_UPDATE_PASSCODE) + 1:][:4]
    with passcode_lock:
        passcode = new_passcode
    logger.debug("New passcode is %s", new_passcode)
    run_speak(POLLY_PROMPT_SENDING_SMS, config["host_region"])


def cmd_handler_ask_secret(config):
    """Handle ASK SECRET command."""
    logger.info("Ask Secret")
    with passcode_lock:
        current_passcode = passcode
    run_passcode(
        current_passcode,
        POLLY_PROMPT_ASK_SECRET,
        POLLY_PROMPT_ALLOW_ACCESS,
        POLLY_PROMPT_DENY_ACCESS,
        config["host_region"],
    )


def cmd_handler_allow_access(config):
    """Handle ALLOW ACCESS command."""
    logger.info("Allow Access")
    run_speak(POLLY_PROMPT_ALLOW_ACCESS, config["host_region"])


def cmd_handler_deny_access(config):
    """Handle DENY ACCESS command."""
    logger.info("Deny Access")
    run_speak(POLLY_PROMPT_DENY_ACCESS, config["host_region"])


def cmd_handler_sms_failed(config):
    """Handle SMS FAILED command."""
    logger.info("SMS Failed")
    run_speak(POLLY_PROMPT_SMS_FAILED, config["host_region"])


# ---------------------------------------------------------------------------
# MQTT Callbacks
# ---------------------------------------------------------------------------

def on_message_received(topic, payload, dup, qos, retain, config, **kwargs):
    """Callback for messages received on subscribed topics."""
    message = payload.decode("utf-8")
    logger.info("Received message on %s: %s", topic, message)

    if message == CMD_CAPTURE_PHOTO:
        cmd_handler_capture_photo(config)
    elif message == CMD_FR_FAILURE:
        cmd_handler_fr_failure(config)
    elif message.startswith(CMD_UPDATE_PASSCODE):
        cmd_handler_update_passcode(message, config)
    elif message == CMD_ASK_SECRET:
        cmd_handler_ask_secret(config)
    elif message == CMD_ALLOW_ACCESS:
        cmd_handler_allow_access(config)
    elif message == CMD_DENY_ACCESS:
        cmd_handler_deny_access(config)
    elif message == CMD_SMS_FAILED:
        cmd_handler_sms_failed(config)
    else:
        logger.warning("Unknown command: %s", message)


def on_connection_interrupted(connection, error, **kwargs):
    """Callback when connection is interrupted."""
    logger.warning("Connection interrupted: %s", error)


def on_connection_resumed(connection, return_code, session_present, **kwargs):
    """Callback when connection is resumed (auto-reconnect)."""
    logger.info("Connection resumed. Return code: %s, Session present: %s",
                return_code, session_present)


# ---------------------------------------------------------------------------
# Startup - publish IP addresses
# ---------------------------------------------------------------------------

def on_startup(mqtt_connection, config):
    """Announce startup: publish IP addresses and speak ready prompt."""
    logger.info("Announcing startup")
    run_speak(POLLY_PROMPT_READY, config["host_region"])

    # Publish wlan0 IP
    wlan0_ip = get_self_ip("wlan0")
    if wlan0_ip:
        msg = f"wlan0 IP address is {wlan0_ip}"
        logger.info(msg)
        mqtt_connection.publish(
            topic=TOPIC_LOCKS_IP,
            payload=msg,
            qos=mqtt.QoS.AT_LEAST_ONCE,
        )

    # Publish eth0 IP
    eth0_ip = get_self_ip("eth0")
    if eth0_ip:
        msg = f"eth0 IP address is {eth0_ip}"
        logger.info(msg)
        mqtt_connection.publish(
            topic=TOPIC_LOCKS_IP,
            payload=msg,
            qos=mqtt.QoS.AT_LEAST_ONCE,
        )


# ---------------------------------------------------------------------------
# Signal Handling
# ---------------------------------------------------------------------------

def signal_handler(signum, frame):
    """Handle SIGINT/SIGTERM for graceful shutdown."""
    logger.info("Received signal %d, shutting down...", signum)
    shutdown_event.set()


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    """Main entry point for aidoorlock."""
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Load configuration
    config = load_config()
    logger.info("Configuration loaded successfully")

    # Set up AWS CRT event loop
    event_loop_group = io.EventLoopGroup(1)
    host_resolver = io.DefaultHostResolver(event_loop_group)
    client_bootstrap = io.ClientBootstrap(event_loop_group, host_resolver)

    # Build MQTT connection with mutual TLS
    mqtt_connection = mqtt_connection_builder.mtls_from_path(
        endpoint=config["mqtt_host"],
        port=config.get("mqtt_port", 8883),
        cert_filepath=config["cert_path"],
        pri_key_filepath=config["key_path"],
        ca_filepath=config["root_ca_path"],
        client_bootstrap=client_bootstrap,
        client_id=config["client_id"],
        clean_session=True,
        keep_alive_secs=30,
        on_connection_interrupted=on_connection_interrupted,
        on_connection_resumed=on_connection_resumed,
    )

    logger.info("Connecting to %s:%d...", config["mqtt_host"],
                config.get("mqtt_port", 8883))
    connect_future = mqtt_connection.connect()
    connect_future.result()
    logger.info("Connected!")

    # Subscribe to commands topic
    logger.info("Subscribing to topic: %s", TOPIC_LOCKS_CMD)
    subscribe_future, packet_id = mqtt_connection.subscribe(
        topic=TOPIC_LOCKS_CMD,
        qos=mqtt.QoS.AT_LEAST_ONCE,
        callback=lambda topic, payload, dup, qos, retain, **kwargs: on_message_received(
            topic, payload, dup, qos, retain, config, **kwargs
        ),
    )
    subscribe_result = subscribe_future.result()
    logger.info("Subscribed with QoS: %s", str(subscribe_result["qos"]))

    # Announce startup
    on_startup(mqtt_connection, config)

    # Main loop - wait for shutdown signal
    logger.info("Door lock is running. Press Ctrl+C to exit.")
    while not shutdown_event.is_set():
        time.sleep(1)

    # Graceful disconnect
    logger.info("Disconnecting...")
    disconnect_future = mqtt_connection.disconnect()
    disconnect_future.result()
    logger.info("Disconnected. Goodbye!")


if __name__ == "__main__":
    main()
